### PR TITLE
Customisable columns in GUI Modlist

### LINKED ISSUE #2687: Feature Request: Customise columns in GUI Modlist
Background
----------
Several changes have added columns to the GUI modlist, providing more information for users. But some users have limited screen real estate and may not be interested in seeing all the new columns (perhaps even some of the older, original columns).

Proposal
---------
CKAN should have a way to customise the columns that appear in the modlist, turning specific columns on or off. We could then make use of this feature to activate certain columns only when there is data (such as the Update/Replace columns)

### PR DESCRIPTION
**Pull request by DasSkelett**
**This PR adds the option to hide columns in the modlist.** as asked by @politas in #2687.

This happens via a right click context menu on the header row of the modlist:
ckan1

The columns are hidden with `column.Visibility = false`, which seems to work perfectly fine in my testing. This doesn't need a reload of the modlist or anything, so it's fast.
On every change of the visibility of a column it is saved to the GUIconfig, which means it's persistend across ckan restarts.

The `Update` and the `Replace` checkbox columns are hidden by default, and only shown if an update / a replacement is available. I could only test updates yet, but replacements should work the same.
Once all updates / replacements have been done, the columns are hidden again.
This all happens in `UpdateModsList()`.

Furthermore, the `Installed version` and the `Install date` columns are hidden with the `Not Installed` filter setting selected, there's no sense in showing them. This happens *without* a GUIconfig change, so once the filter is changed again, those two columns reappear (or stay hidden) based on the previous setting.

many columns hidden updates available

Closes #2687 

---
Small sidefix: if you right-clicked on the modlist header bar before, the context menu for the mod rows opened, despite it shouldn't. It's fixed now by not assigning `ModList.ContextMenuStrip = ModListContextMenuStrip`. Instead, the new column selection menu opens, obviously.

### CODE REVIEW

**Comment by HebaruSan** [comment]
Let's remove the update/replace columns from the menu since we're managing them automatically.

**Comment by DasSkelett** [comment]
I'm going to set the `Installed` checkbox column (the first one) to always visible then true.
Else the code will get a bit messy, because I access the columns via their index (which is taken from the index of the clicked button).

**Comment by DasSkelett** [comment]
Removed the buttons for the first three columns in the context menu.
Beware: If you've already run the previous commit, make sure to either delete the whole GUIConfig.xml or only the `VisibleColumns` array, because I shortened the array, and "using" the old one will be buggy.

**Comment by HebaruSan** [comment]
This looks pretty good. Created DasSkelett/CKAN#3 with some code style and UI ideas.

**Comment by DasSkelett** [comment]
For the record: @HebaruSan's suggestions are merged and included. Thanks :)

File: GUI/MainModList.cs [hunk 1]
```diff
@@ -273,6 +274,14 @@ private void _UpdateModsList(bool repo_updated, IEnumerable<ModChange> mc)
 
             UpdateFilters(this);
 
+            // Hide update and replacement columns if not needed.
+            // Write it to the configuration, else they are hidden agian after a filter change.
+            // After the update / replacement, they are hidden again.
+            ModList.Columns[1].Visible = has_any_updates;
+            configuration.VisibleColumns[1] = has_any_updates;
```
Comment (HebaruSan) [inline GUI/MainModList.cs]: Configuration probably should be modified here if we're no longer including update/replace in the list.

**Comment by HebaruSan** [comment]
DasSkelett/CKAN#4 created with a suggestion for the config.

**Pull request merged**